spec.bs

```diff
@@ -447,6 +448,83 @@ To <dfn>build an interest group permissions url</dfn> given a [=origin=] |ownerO
 
 </div>
 
+<h3 id="interest-group-storage-maintenance">Interest Group Storage Maintenance</h3>
+
+There is a job that periodically [=performs storage maintenance=] on the user agent's
+[=interest group set=]. It performs operations such as [=list/removing=] expired or excess
+[=interest groups=]. It's required that [=interest group set=] has no more than:
+  * 2000 [=interest group/owners=];
+  * 1000 [=regular interest groups=] per [=interest group/owner=];
+  * 20000 [=negative interest groups=] per [=interest group/owner=];
+  * <code>10\*1024\*1024</code> total [=interest group/estimated size=] of [=interest groups=] per
+    [=interest group/owner=].
+
+<div algorithm='maintenance'>
+To <dfn>perform storage maintenance</dfn>:
+
+1. Let |ownersAndExpiry| be a new [=ordered map=] whose [=map/keys=] are [=origins=]
+  [=interest group/owner=] and [=map/values=] are [=moments=] [=interest group/expiry=].
+1. [=list/For each=] |ig| of the user agent's [=interest group set=]:
+  1. Let |owner| be |ig|'s [=interest group/owner=].
+  1. If |ownersAndExpiry|[|owner|] [=map/exists=], then [=map/set=] |ownersAndExpiry|[|owner|] to
+    |ig|'s [=interest group/expiry=] if it comes after |ownersAndExpiry|[|owner|].
+  1. Otherwise, [=map/set=] |ownersAndExpiry|[|owner|] to |ig|'s [=interest group/expiry=].
+1. If [=map/size=] of |ownersAndExpiry| |i| &gt; 2000, then set |ownersAndExpiry| to
+  |ownersAndExpiry| [=map/sorted in descending order=] with |a| being less than |b| if |a|'s
+  [=map/value=] comes before |b|'s [=map/value=].
+
+  Note: Sorting owners based on their expiry in descending order so that if interest groups of some
+    owners need to be removed, the interest groups of owners expiring soonest are the ones to be
+    removed first.
+
+1. Let |owners| be the [=map/get the keys|keys=] of |ownersAndExpiry|.
+1. [=list/For each=] |i| in [=the range=] from 0 to the [=set/size=] of |owners|, exclusive:
+  1. If |i| &ge; 2000, then [=list/remove=] [=interest groups=] from the user agent's
+    [=interest group set=] whose [=interest group/owner=] is |owners|[|i|], and [=iteration/continue=].
+  1. Let |regularIgs| be a [=list=] of [=regular interest groups=] in the user agent's
+    [=interest group set=] whose [=interest group/owner=] is |owners|[|i|].
+  1. If [=list/size=] of |regularIgs| &gt; 1000, then [=clear excess interest groups=] with
+    |regularIgs|, |owners|[|i|], and 1000.
+  1. Let |negativeIgs| be a [=list=] of [=negative interest groups=] in the user agent's
+    [=interest group set=] whose [=interest group/owner=] is |owners|[|i|].
+  1. If [=list/size=] of |regularIgs| &gt; 20000, then [=Clear excess interest groups=] with
+    |negativeIgs|, |owners|[|i|], and 20000.
+1. TODO: remove expired IGs.

```

Comment:
I think all TODOs from here on out should be a follow-up bug with a `<span class=XXX>` or `<p class=XXX>`, since this is a real specification that has a shipping implementation. Please file follow-up bugs for TODOs like this, and add a description over there of how the final design should look, and what the goals are specifically.